java-util/src/main/java/io/druid/java/util/common/io/NativeIO.java

```diff
@@ -0,0 +1,171 @@
+ * under the License.
+ */
+
+package io.druid.java.util.common.io;
+
+import java.io.File;
+import java.io.FileDescriptor;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.RandomAccessFile;
+import java.lang.reflect.Field;
+
+import com.sun.jna.LastErrorException;
+import com.sun.jna.Native;
+import com.sun.jna.Platform;
+import io.druid.java.util.common.logger.Logger;
+
+/**
+ * Native I/O operations in order to minimize cache impact.
+ */
+public class NativeIO
+{
+  private static final Logger log = new Logger(NativeIO.class);
+
+  private static final int POSIX_FADV_DONTNEED = 4; /* fadvise.h */
+  private static Field field;
+
+  private static volatile boolean initialized = false;
+  private static volatile boolean fadvisePossible = true;
+
+  static {
+    field = getFieldByReflection(FileDescriptor.class, "fd");
+
+    try {
+      Native.register(Platform.C_LIBRARY_NAME);
+      initialized = true;
+    }
+    catch (NoClassDefFoundError e) {
+      log.info("JNA not found. Native methods will be disabled.");
+    }
+    catch (UnsatisfiedLinkError e) {
+      log.info("Unable to link C library. Native methods will be disabled.");
+    }
+    catch (NoSuchMethodError e) {
+      log.warn("Obsolete version of JNA present; unable to register C library");
+    }
+  }
+
+  private static native int posix_fadvise(int fd, long offset, long len, int flag) throws LastErrorException;
+
+  private NativeIO() {}
+
+  private static Field getFieldByReflection(Class cls, String fieldName)
+  {
+    Field field = null;
+
+    try {
+      field = cls.getDeclaredField(fieldName);
+      field.setAccessible(true);
+    }
+    catch (Exception e) {
+      log.warn("Unable to read [%s] field from [%s]", fieldName, cls.getName());
+    }
+
+    return field;
+  }
+  /**
+   * Get system file descriptor (int) from FileDescriptor object.
+   * @param descriptor - FileDescriptor object to get fd from
+   * @return file descriptor, -1 or error
+   */
+  public static int getfd(FileDescriptor descriptor)
+  {
+    try {
+      return field.getInt(descriptor);
+    }
+    catch (IllegalArgumentException | IllegalAccessException e) {
+      log.warn("Unable to read fd field from java.io.FileDescriptor");
+    }
+
+    return -1;
+  }
+
+  /**
+   * Remove pages from the file system page cache when they wont
+   * be accessed again
+   *
+   * @param fd     The file descriptor of the source file.
+   * @param offset The offset within the file.
+   * @param len    The length to be flushed.
+   */
+
+  public static void trySkipCache(int fd, long offset, long len)
+  {
+    if (!initialized || !fadvisePossible || fd < 0) {
+      return;
+    }
+    try {
+      // we ignore the return value as this is just best effort to avoid the cache
+      posix_fadvise(fd, offset, len, POSIX_FADV_DONTNEED);
+    }
+    catch (UnsupportedOperationException uoe) {
+      log.warn(uoe, "posix_fadvise is not supported");
```

Comment:
(minor) suggest just making these `info` since they are recoverable

Reply:
Can we leave this at warn? It will only be logged once anyway.